README.md
# Sunrise-Sunset JS

[![npm](https://img.shields.io/npm/v/sunrise-sunset-api.svg)](https://www.npmjs.com/package/sunrise-sunset-api)
[![license](https://img.shields.io/npm/l/sunrise-sunset-api.svg)](https://github.com/dguo/sunrise-sunset-api/blob/main/LICENSE.txt)
[![bundle size](https://img.shields.io/bundlephobia/minzip/sunrise-sunset-api)](https://bundlephobia.com/result?p=sunrise-sunset-api)
[![CI status](https://github.com/dguo/sunrise-sunset-js/workflows/CI/badge.svg)](https://github.com/dguo/sunrise-sunset-js/actions?query=branch%3Amain)
[![test coverage](https://codecov.io/gh/dguo/sunrise-sunset-js/branch/main/graph/badge.svg)](https://codecov.io/gh/dguo/sunrise-sunset-js)
[![known vulnerabilities](https://snyk.io/test/github/dguo/sunrise-sunset-js/badge.svg?targetFile=package.json)](https://snyk.io/test/github/dguo/sunrise-sunset-js?targetFile=package.json)

JavaScript client for the [Sunrise-Sunset API](https://sunrise-sunset.org/api).
This client works in browser and Node.js environments.

## Installation

```sh
npm install --save sunrise-sunset-api
```

or

```sh
yarn add sunrise-sunset-api
```

## Usage

```js
import {getSunriseSunsetInfo} from "sunrise-sunset-api";

(async function main() {
    const response = await getSunriseSunsetInfo({
        latitude: 36.72016,
        longitude: -44.42034,
    });
})();
```

The response is in this format. Note that all times are in
[UTC](https://en.wikipedia.org/wiki/Coordinated_Universal_Time).

```js
{
    sunrise: "8:02:42 AM",
    sunset: "9:46:40 PM",
    solarNoon: "2:54:41 PM",
    dayLength: "13:43:58",
    civilTwilightBegin: "7:34:56 AM",
    civilTwilightEnd: "10:14:26 PM",
    nauticalTwilightBegin: "7:01:18 AM",
    nauticalTwilightEnd: "10:48:04 PM",
    astronomicalTwilightBegin: "6:25:35 AM",
    astronomicalTwilightEnd: "11:23:47 PM"
}
```

You can get the response for a specific date (in `YYYY-MM-DD` format) with the
`date` option. If you don't pass a date, you'll get the response for whatever
the API thinks is the current date.

```js
const response = await getSunriseSunsetInfo({
    latitude: 36.72016,
    longitude: -44.42034,
    date: "2021-10-31"
});
```

You can turn off formatting with the `formatted` option.

```js
const response = await getSunriseSunsetInfo({
    latitude: 36.72016,
    longitude: -44.42034,
    formatted: false
});
```

Which makes the info come back in this format: `dayLength` is an integer number
of seconds, and the times are in [ISO
8601](https://en.wikipedia.org/wiki/ISO_8601) format.

```js
{
    sunrise: "2021-05-01T08:02:42+00:00",
    sunset: "2021-05-01T21:46:40+00:00",
    solarNoon: "2021-05-01T14:54:41+00:00",
    dayLength: 49438,
    civilTwilightBegin: "2021-05-01T07:34:56+00:00",
    civilTwilightEnd: "2021-05-01T22:14:26+00:00",
    nauticalTwilightBegin: "2021-05-01T07:01:18+00:00",
    nauticalTwilightEnd: "2021-05-01T22:48:04+00:00",
    astronomicalTwilightBegin: "2021-05-01T06:25:35+00:00",
    astronomicalTwilightEnd: "2021-05-01T23:23:47+00:00"
}
```

## Attribution

Don't forget to properly attribute the Sunrise-Sunset API with a link to
their site.

## Ky

This client uses [Ky](https://github.com/sindresorhus/ky) to make the API
request. Be aware that Ky has a built-in
[retry](https://github.com/sindresorhus/ky#retry) mechanism for failed requests.

Ky is built on the [Fetch
API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API), and here's an
example of how you might handle a Fetch error.

```js
import ky from "ky-universal";

try {
    const response = await getSunriseSunsetInfo({
        latitude: 36.72016,
        longitude: -44.42034,
    });
} catch (error) {
    if (error instanceof ky.HTTPError && error.response.status === 400) {
        // Get the actual body that the API responded with
        console.error(await error.response.json());
    }
    console.error(error);
}
```

## API

### getSunriseSunsetInfo(options)

#### options

Type: `object`

##### latitude

Type: `number`

Should be a value between -90 and 90 (inclusive).

##### longitude

Type: `number`

Should be a value between -180 and 180 (inclusive).

##### date

Type: `string | null`\
Default: `current date (according to the API)`

Pass this to get the info for a particular date. The value should be in a
`YYYY-MM-DD` format.

##### formatted

Type: `boolean`\
Default: `true`

Formatted responses have times like `8:02:42 AM` and a `dayLength` like
`13:43:58` (`HH:MM:SS` format). Unformatted responses have times like
`2021-05-01T08:02:42+00:00` (ISO 8601 format) and a `dayLength` like `49438`
(integer number of seconds).

##### apiUrl

Type: `string`\
Default: `https://api.sunrise-sunset.org/json`

Pass this to override the API endpoint if it changes, and I am slow to update the
code.

##### kyOptions

Type: `object`\
Default: `{}`

Pass this to override any [Ky
options](https://github.com/sindresorhus/ky/blob/main/readme.md#options).

##### useMocks

Type: `boolean`\
Default: `false`

Turn on [mock mode](#Mocks).

## Getting the Latitude and Longitude

In the browser, you can use the [Geolocation
API](https://developer.mozilla.org/en-US/docs/Web/API/Geolocation_API) to get
the user's latitude and longitude, assuming the user gives permission.

In a Node.js environment, you can look up the coordinates based on an IP
address. There are several APIs for this, like [IPinfo](https://ipinfo.io/),
[ipstack](https://ipstack.com/), and [ipapi](https://ipapi.co/).

## Types

This library is written in [TypeScript](https://www.typescriptlang.org/) and
exports some types that you might find useful.

```ts
import {
    SunriseSunsetOptions,
    UnformattedSunriseSunsetResponse,
    FormattedSunriseSunsetResponse,
    SunriseSunsetResponse
} from "sunrise-sunset-js";
```

`SunriseSunsetOptions` describes the object that is passed to
`getSunriseSunsetInfo`.

`SunriseSunsetResponse` is just a union of the unformatted and formatted
responses.

## Mocks

For automated tests or for any other situation where you don't want to make
actual network requests, you can turn on mock mode, which will make the function
return hardcoded responses.

```js
const response = await getSunriseSunsetInfo({
    latitude: 36.72016,
    longitude: -44.42034,
    useMocks: true
});
```

You can also import the mock responses.

```js
import {
    MOCK_FORMATTED_RESPONSE,
    MOCK_UNFORMATTED_RESPONSE
} from "sunrise-sunset-js";

mockFunction.mockResolvedValue(MOCK_FORMATTED_RESPONSE);
```

Of course, you can always come up with your own mocks or overwrite specific
fields as necessary.

```js
const mockResponse = {
    ...MOCK_FORMATTED_RESPONSE,
    dayLength: "12:00:00"
}
```

## License

MIT


CHANGELOG.md
# Changelog

This project attempts to adhere to [Semantic Versioning](http://semver.org).

## [Unreleased]

## [0.2.1] - (2021-05-02)

### Fixed

- Fixed the homepage and repository links in `package.json`.

## [0.2.0] - (2021-05-01)

### Added

- Exported the `UnformattedSunriseSunsetResponse`,
  `FormattedSunriseSunsetResponse`, and `SunriseSunsetResponse` types.

### Changed

- Renamed the `SunriseSunsetRequest` type to `SunriseSunsetOptions`.

### Removed

- Removed the option to get keys in camel case format. This is the default
  format now, with no way to get keys in snake case format. This is to adhere
  to the convention for JavaScript. If you need snake case keys, you can convert
  them yourself.

### Fixed

- Fixed mock mode incorrectly returning an unformatted response when the
  formatting mode isn't specified.

## [0.1.0] - (2021-03-27)

### Added

- Initial implementation

[unreleased]: https://github.com/dguo/sunrise-sunset-js/compare/v0.2.1...HEAD
[0.2.1]: https://github.com/dguo/sunrise-sunset-js/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com/dguo/sunrise-sunset-js/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/dguo/sunrise-sunset-js/releases/tag/v0.1.0


source/index.test.ts
import fetchMock from "fetch-mock-jest";

import {getSunriseSunsetInfo} from "./index";
import * as mocks from "./mocks";

fetchMock
    .mock(
        (url) => !url.includes("formatted=") || url.includes("formatted=1"),
        mocks.MOCK_RAW_FORMATTED_RESPONSE
    )
    .mock(
        (url) => url.includes("formatted=0"),
        mocks.MOCK_RAW_UNFORMATTED_RESPONSE
    );

test("Successful unformatted response", async () => {
    const info = await getSunriseSunsetInfo({
        latitude: 34.5,
        longitude: 88.88,
        date: "2020-03-01",
        formatted: false,
    });

    expect(info).toEqual(mocks.MOCK_UNFORMATTED_RESPONSE);
});

test("Successful formatted response", async () => {
    const info = await getSunriseSunsetInfo({
        latitude: 34.5,
        longitude: 88.88,
        formatted: true,
    });

    expect(info).toEqual(mocks.MOCK_FORMATTED_RESPONSE);
});

test("Can override the API URL", async () => {
    const newUrl = "https://example.com";

    await getSunriseSunsetInfo({
        latitude: 34.5,
        longitude: 88.88,
        apiUrl: newUrl,
    });

    expect(fetchMock.lastUrl()).toMatch(newUrl);
});

test("Can override Ky options", async () => {
    await getSunriseSunsetInfo({
        latitude: 34.5,
        longitude: 88.88,
        kyOptions: {
            searchParams: {
                foo: "bar",
            },
        },
    });

    expect(fetchMock.lastUrl()).toMatch("foo=bar");
});

test("Mock mode works when there's no explicit formatting mode", async () => {
    const info = await getSunriseSunsetInfo({
        latitude: 34.5,
        longitude: 88.88,
        useMocks: true,
    });

    expect(info).toEqual(mocks.MOCK_FORMATTED_RESPONSE);
});

test("Mock mode works when explicitly turning formatted mode on", async () => {
    const info = await getSunriseSunsetInfo({
        latitude: 34.5,
        longitude: 88.88,
        useMocks: true,
        formatted: true,
    });

    expect(info).toEqual(mocks.MOCK_FORMATTED_RESPONSE);
});

test("Mock mode works when explicitly turning formatted mode off", async () => {
    const info = await getSunriseSunsetInfo({
        latitude: 34.5,
        longitude: 88.88,
        useMocks: true,
        formatted: false,
    });

    expect(info).toEqual(mocks.MOCK_UNFORMATTED_RESPONSE);
});

test.each([100, "foo", -90.1, NaN, null])(
    "Errors for an invalid latitude: %p",
    async (invalidLatitude) => {
        await expect(
            getSunriseSunsetInfo({
                latitude: invalidLatitude as number,
                longitude: 88.88,
            })
        ).rejects.toThrowError(/latitude/i);
    }
);

test.each([180.1, "bar", -190, NaN, null])(
    "Errors for a bad longitude: %p",
    async (invalidLongitude) => {
        await expect(
            getSunriseSunsetInfo({
                latitude: 50,
                longitude: invalidLongitude as number,
            })
        ).rejects.toThrowError(/longitude/i);
    }
);

test.each([42, true])("Errors for a bad date: %p", async (invalidDate) => {
    await expect(
        getSunriseSunsetInfo({
            latitude: 50,
            longitude: 70,
            date: (invalidDate as unknown) as string,
        })
    ).rejects.toThrowError(/date/i);
});

test.each([{formatted: true}, {formatted: false}])(
    "Mock mode works with: %p",
    async (options) => {
        fetchMock.resetHistory();

        await getSunriseSunsetInfo({
            latitude: 23.46,
            longitude: -50,
            formatted: options.formatted,
            useMocks: true,
        });

        expect(fetchMock.calls().length).toEqual(0);
    }
);


source/index.ts
import ky from "ky-universal";
import {Options as KyOptions} from "ky";
import camelCaseKeys from "camelcase-keys";

import {MOCK_FORMATTED_RESPONSE, MOCK_UNFORMATTED_RESPONSE} from "./mocks";

export {MOCK_FORMATTED_RESPONSE, MOCK_UNFORMATTED_RESPONSE};

interface BaseSunriseSunsetOptions {
    latitude: number;
    longitude: number;
    date?: string | null;
    apiUrl?: string;
    kyOptions?: KyOptions;
    useMocks?: boolean;
}

export interface SunriseSunsetOptions extends BaseSunriseSunsetOptions {
    formatted?: boolean;
}

interface BaseSunriseSunsetResponse {
    sunrise: string;
    sunset: string;
    solarNoon: string;
    civilTwilightBegin: string;
    civilTwilightEnd: string;
    nauticalTwilightBegin: string;
    nauticalTwilightEnd: string;
    astronomicalTwilightBegin: string;
    astronomicalTwilightEnd: string;
}

export interface UnformattedSunriseSunsetResponse
    extends BaseSunriseSunsetResponse {
    dayLength: string;
}

export interface FormattedSunriseSunsetResponse
    extends BaseSunriseSunsetResponse {
    dayLength: number;
}

export type SunriseSunsetResponse =
    | FormattedSunriseSunsetResponse
    | UnformattedSunriseSunsetResponse;

export async function getSunriseSunsetInfo(
    options: BaseSunriseSunsetOptions & {formatted: false}
): Promise<UnformattedSunriseSunsetResponse>;

export async function getSunriseSunsetInfo(
    options: BaseSunriseSunsetOptions & {formatted?: true}
): Promise<FormattedSunriseSunsetResponse>;

export async function getSunriseSunsetInfo(
    options: BaseSunriseSunsetOptions & {formatted: boolean}
): Promise<SunriseSunsetResponse>;

/* We need to duplicate the implementation signature because only the overloads
   contribute to the final shape of the function signature. See:
   https://github.com/Microsoft/TypeScript/wiki/FAQ#why-am-i-getting-supplied-parameters-do-not-match-any-signature-error
   */
export async function getSunriseSunsetInfo(
    options: SunriseSunsetOptions
): Promise<SunriseSunsetResponse>;

export async function getSunriseSunsetInfo(
    options: SunriseSunsetOptions
): Promise<SunriseSunsetResponse> {
    if (!options.latitude) {
        throw new Error("Latitude is a required parameter");
    }
    if (!options.longitude) {
        throw new Error("Longitude is a required parameter");
    }
    if (
        typeof options.latitude !== "number" ||
        !isFinite(options.latitude) ||
        Math.abs(options.latitude) > 90
    ) {
        throw new Error(
            "Latitude must be a number between -90 and 90 (inclusive)"
        );
    }
    if (
        typeof options.longitude !== "number" ||
        !isFinite(options.longitude) ||
        Math.abs(options.longitude) > 180
    ) {
        throw new Error(
            "Longitude must be a number between -180 and 180 (inclusive)"
        );
    }
    if (
        typeof options.date !== "undefined" &&
        options.date !== null &&
        typeof options.date !== "string"
    ) {
        throw new Error("Invalid date");
    }

    if (options.useMocks) {
        if (typeof options.formatted === "undefined" || options.formatted) {
            return MOCK_FORMATTED_RESPONSE;
        } else {
            return MOCK_UNFORMATTED_RESPONSE;
        }
    }

    const response = await ky(
        options.apiUrl || "https://api.sunrise-sunset.org/json",
        {
            method: "get",
            searchParams: {
                lat: options.latitude,
                lng: options.longitude,
                ...(typeof options.date === "string" && {date: options.date}),
                ...(typeof options.formatted === "boolean" && {
                    formatted: options.formatted ? 1 : 0,
                }),
            },
            ...options.kyOptions,
        }
    );

    const body = await response.json();

    return camelCaseKeys(body.results);
}


source/mocks.ts
export const MOCK_RAW_FORMATTED_RESPONSE = {
    results: {
        sunrise: "7:27:02 AM",
        sunset: "5:05:55 PM",
        solar_noon: "12:16:28 PM",
        day_length: "9:38:53",
        civil_twilight_begin: "6:58:14 AM",
        civil_twilight_end: "5:34:43 PM",
        nautical_twilight_begin: "6:25:47 AM",
        nautical_twilight_end: "6:07:10 PM",
        astronomical_twilight_begin: "5:54:14 AM",
        astronomical_twilight_end: "6:38:43 PM",
    },
    status: "OK",
};

export const MOCK_FORMATTED_RESPONSE = {
    sunrise: "7:27:02 AM",
    sunset: "5:05:55 PM",
    solarNoon: "12:16:28 PM",
    dayLength: "9:38:53",
    civilTwilightBegin: "6:58:14 AM",
    civilTwilightEnd: "5:34:43 PM",
    nauticalTwilightBegin: "6:25:47 AM",
    nauticalTwilightEnd: "6:07:10 PM",
    astronomicalTwilightBegin: "5:54:14 AM",
    astronomicalTwilightEnd: "6:38:43 PM",
};

export const MOCK_RAW_UNFORMATTED_RESPONSE = {
    results: {
        sunrise: "2015-05-21T05:05:35+00:00",
        sunset: "2015-05-21T19:22:59+00:00",
        solar_noon: "2015-05-21T12:14:17+00:00",
        day_length: 51444,
        civil_twilight_begin: "2015-05-21T04:36:17+00:00",
        civil_twilight_end: "2015-05-21T19:52:17+00:00",
        nautical_twilight_begin: "2015-05-21T04:00:13+00:00",
        nautical_twilight_end: "2015-05-21T20:28:21+00:00",
        astronomical_twilight_begin: "2015-05-21T03:20:49+00:00",
        astronomical_twilight_end: "2015-05-21T21:07:45+00:00",
    },
    status: "OK",
};

export const MOCK_UNFORMATTED_RESPONSE = {
    sunrise: "2015-05-21T05:05:35+00:00",
    sunset: "2015-05-21T19:22:59+00:00",
    solarNoon: "2015-05-21T12:14:17+00:00",
    dayLength: 51444,
    civilTwilightBegin: "2015-05-21T04:36:17+00:00",
    civilTwilightEnd: "2015-05-21T19:52:17+00:00",
    nauticalTwilightBegin: "2015-05-21T04:00:13+00:00",
    nauticalTwilightEnd: "2015-05-21T20:28:21+00:00",
    astronomicalTwilightBegin: "2015-05-21T03:20:49+00:00",
    astronomicalTwilightEnd: "2015-05-21T21:07:45+00:00",
};
